src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala

```diff
@@ -71,36 +73,68 @@ class TaskStatusUpdateProcessorImpl @Inject() (
       case Some(instance) if taskIsUnknown(instance, taskId) =>
         if (killWhenUnknown(taskCondition)) {
           killUnknownTaskTimeMetric {
-            logger.warn(s"Kill ${taskId} because it's unknown to marathon. " +
+            logger.warn(s"Kill $taskId because it's unknown to marathon. " +
               s"The related instance ${instance.instanceId} is associated with ${instance.tasksMap.keys}")
             Future.successful(killService.killUnknownTask(taskId, KillReason.NotInSync))
           }
         }
+        checkDelay(instance, status)
         acknowledge(status)
 
       case Some(instance) =>
         // TODO(PODS): we might as well pass the taskCondition here
-        instanceTracker.updateStatus(instance, status, now).flatMap(_ => acknowledge(status))
+        instanceTracker.updateStatus(instance, status, now).flatMap{ _ =>
+          checkDelay(instance, status)
+          acknowledge(status)
+        }
 
       case None if terminalUnknown(taskCondition) =>
-        logger.warn(s"Received terminal status update for unknown ${taskId}")
+        logger.warn(s"Received terminal status update for unknown $taskId")
         eventStream.publish(UnknownInstanceTerminated(taskId.instanceId, taskId.runSpecId, taskCondition))
         acknowledge(status)
 
       case None if killWhenUnknown(taskCondition) =>
         killUnknownTaskTimeMetric {
-          logger.warn(s"Kill unknown ${taskId}")
+          logger.warn(s"Kill unknown $taskId")
           killService.killUnknownTask(taskId, KillReason.Unknown)
           acknowledge(status)
         }
 
-      case maybeTask: Option[Instance] =>
-        val taskStr = taskKnownOrNotStr(maybeTask)
+      case maybeInstance: Option[Instance] =>
+        val taskStr = taskKnownOrNotStr(maybeInstance)
         logger.info(s"Ignoring ${status.getState} update for $taskStr $taskId")
         acknowledge(status)
     }
   }
 
+  /**
+    * @return true, if the received status indicates that the agent it comes from is being drained.
+    *
+    * The only relevant states here are TASK_KILLED and TASK_GONE_BY_OPERATOR. See the design doc
+    * https://docs.google.com/document/d/1w3O80NFE6m52XNMv7EdXSO-1NebEs8opA8VZPG1tW0Y/edit#
+    * for more info
+    */
+  def agentDraining(status: MesosProtos.TaskStatus): Boolean = {
+    val relevantState = status.hasState &&
+      (status.getState == MesosProtos.TaskState.TASK_KILLED || status.getState == MesosProtos.TaskState.TASK_GONE_BY_OPERATOR)
+    val isDraining = status.hasReason && status.getReason == MesosProtos.TaskStatus.Reason.REASON_SLAVE_DRAINING
+
+    relevantState && isDraining
+  }
+
+  /**
+    * When receiving task statuses from draining agents, we reset any delay
+    * that might exist for the related runSpec. This is to speed up the
+    * process of replacing these tasks.
+    */
+  def checkDelay(instance: Instance, status: MesosProtos.TaskStatus): Unit = {
```

Comment:
I generally don't like that method called `check` is doing some state modification (resetting delay)

Reply:
in general it feels pretty hacky to have this logic inside class processing status updates 😛

Reply:
agree!

Reply:
Thanks for that feedback folks – I initially looked into `NotifyRateLimiterStepImpl` but dropped that because it seemed it would only complicate things further. I looked into it again and realized there meanwhile totally unnecessary app/pod fetching I could remove from that step – looks much simpler now :D